options/commonOptions.go

```diff
@@ -0,0 +1,666 @@
+package options
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	dtlsServer "github.com/plgd-dev/go-coap/v3/dtls/server"
+	"github.com/plgd-dev/go-coap/v3/message/pool"
+	"github.com/plgd-dev/go-coap/v3/mux"
+	"github.com/plgd-dev/go-coap/v3/net/blockwise"
+	"github.com/plgd-dev/go-coap/v3/net/client"
+	"github.com/plgd-dev/go-coap/v3/net/monitor/inactivity"
+	"github.com/plgd-dev/go-coap/v3/options/config"
+	"github.com/plgd-dev/go-coap/v3/pkg/runner/periodic"
+	tcpClient "github.com/plgd-dev/go-coap/v3/tcp/client"
+	tcpServer "github.com/plgd-dev/go-coap/v3/tcp/server"
+	udpClient "github.com/plgd-dev/go-coap/v3/udp/client"
+	udpServer "github.com/plgd-dev/go-coap/v3/udp/server"
+)
+
+type ErrorFunc = config.ErrorFunc
+
+type GoPoolFunc = config.GoPoolFunc
+
+type Handler interface {
+	tcpClient.HandlerFunc | udpClient.HandlerFunc
+}
+
+// HandlerFuncOpt handler function option.
+type HandlerFuncOpt[H Handler] struct {
+	h H
+}
+
+func panicForInvalidHandlerFunc(t, exp any) {
+	panic(fmt.Errorf("invalid HandlerFunc type %T, expected %T", t, exp))
+}
+
+func (o HandlerFuncOpt[H]) TCPServerApply(cfg *tcpServer.Config) {
+	switch v := any(o.h).(type) {
+	case tcpClient.HandlerFunc:
+		cfg.Handler = v
+	default:
+		var exp tcpClient.HandlerFunc
+		panicForInvalidHandlerFunc(v, exp)
+	}
+}
+
+func (o HandlerFuncOpt[H]) TCPClientApply(cfg *tcpClient.Config) {
+	switch v := any(o.h).(type) {
+	case tcpClient.HandlerFunc:
+		cfg.Handler = v
+	default:
+		var exp tcpClient.HandlerFunc
+		panicForInvalidHandlerFunc(v, exp)
+	}
+}
+
+func (o HandlerFuncOpt[H]) UDPServerApply(cfg *udpServer.Config) {
+	switch v := any(o.h).(type) {
+	case udpClient.HandlerFunc:
+		cfg.Handler = v
+	default:
+		var exp udpClient.HandlerFunc
+		panicForInvalidHandlerFunc(v, exp)
+	}
+}
+
+func (o HandlerFuncOpt[H]) DTLSServerApply(cfg *dtlsServer.Config) {
+	switch v := any(o.h).(type) {
+	case udpClient.HandlerFunc:
+		cfg.Handler = v
+	default:
+		var exp udpClient.HandlerFunc
+		panicForInvalidHandlerFunc(v, exp)
+	}
+}
+
+func (o HandlerFuncOpt[H]) UDPClientApply(cfg *udpClient.Config) {
+	switch v := any(o.h).(type) {
+	case udpClient.HandlerFunc:
+		cfg.Handler = v
+	default:
+		var t udpClient.HandlerFunc
+		panicForInvalidHandlerFunc(v, t)
+	}
+}
+
+// WithHandlerFunc set handle for handling request's.
+func WithHandlerFunc[H Handler](h H) HandlerFuncOpt[H] {
+	return HandlerFuncOpt[H]{h: h}
+}
+
+// HandlerFuncOpt handler function option.
+type MuxHandlerOpt struct {
+	m mux.Handler
+}
+
+func (o MuxHandlerOpt) TCPServerApply(cfg *tcpServer.Config) {
```

Comment:
When they implement generic methods we can get rid of these overload and keep a single method.